src/Streamly/Internal/Data/Array/Mut/Type.hs

```diff
@@ -0,0 +1,535 @@
+
+-- | Write the given element to the given index of the array. Does not check if
+-- the index is out of bounds of the array.
+--
+-- /Pre-release/
+{-# INLINE putIndexUnsafe #-}
+putIndexUnsafe :: forall m a. MonadIO m => Array a -> Int -> a -> m ()
+putIndexUnsafe Array {..} i x =
+    liftInIO
+        $ \s# ->
+              case i + arrStart of
+                  I# n# ->
+                      let s1# = writeArray# arrContents# n# x s#
+                       in (# s1#, () #)
+
+invalidIndex :: String -> Int -> a
+invalidIndex label i =
+    error $ label ++ ": invalid array index " ++ show i
+
+-- | /O(1)/ Write the given element at the given index in the array.
+-- Performs in-place mutation of the array.
+--
+-- >>> putIndex arr ix val = Array.modifyIndex arr ix (const (val, ()))
+--
+-- /Pre-release/
+{-# INLINE putIndex #-}
+putIndex :: MonadIO m => Array a -> Int -> a -> m ()
+putIndex arr@(Array {..}) i x =
+    if i < arrLen && i > 0
+    then putIndexUnsafe arr i x
+    else invalidIndex "putIndex" i
+
+-- | Modify a given index of an array using a modifier function without checking
+-- the bounds.
+--
+-- Unsafe because it does not check the bounds of the array.
+--
+-- /Pre-release/
+modifyIndexUnsafe :: MonadIO m => Array a -> Int -> (a -> (a, b)) -> m b
+modifyIndexUnsafe Array {..} i f = do
+    liftInIO
+        $ \s# ->
+              case i + arrStart of
+                  I# n# ->
+                      case readArray# arrContents# n# s# of
+                          (# s1#, a #) ->
+                              let (a1, b) = f a
+                                  s2# = writeArray# arrContents# n# a1 s1#
+                               in (# s2#, b #)
+
+-- | Modify a given index of an array using a modifier function.
+--
+-- /Pre-release/
+modifyIndex :: MonadIO m => Array a -> Int -> (a -> (a, b)) -> m b
+modifyIndex arr@(Array {..}) i f = do
+    if i < arrLen && i > 0
+    then modifyIndexUnsafe arr i f
+    else invalidIndex "modifyIndex" i
+
+-------------------------------------------------------------------------------
+-- Resizing
+-------------------------------------------------------------------------------
+
+-- | Reallocates the array according to the new size. This is a safe function
+-- that always creates a new array and copies the old array into the new one. If
+-- the reallocated size is less then it results in a truncated version of the
+-- original array.
+--
+realloc :: MonadIO m => Int -> Array a -> m (Array a)
+realloc i arr = do
+    arr1 <- newArray i
+    let !newLen@(I# newLen#) = min i (arrLen arr)
+        !(I# arrS#) = arrStart arr
+        !(I# arr1S#) = arrStart arr1
+        arrC# = arrContents# arr
+        arr1C# = arrContents# arr1
+    liftInIO
+        $ \s# ->
+              let s1# = copyMutableArray# arrC# arrS# arr1C# arr1S# newLen# s#
+               in (# s1#, arr1 {arrLen = newLen, arrFree = i} #)
```

Comment:
Is `arrFree = i` correct? When we are reallocating the free index should remain the same as earlier. `arrFree = i` means nothing is free.

Reply:
1. I shouldn't have renamed it to `arrFree`. I don't know what I was thinking at that point of time.
2. `arrFree` (`arrSpace` initially) is the true length of the array contents itself. The first unacceptable index.